### PR TITLE
Feature/on subscription complete callback

### PR DESCRIPTION
**Pull request by soares-fernando**
Hello guys, how is it going?

I came across using **ariadne** (which is very well written, by the way) for a project in my company. The project consists of a microservice that is going to provide queries and mostly subscriptions for our front-end layer. Within the architecture of the project, I need to control when the client initiates a subscription and when the subscription has ended to turn on and off, respectively, some of the retrieving information mechanisms (task etc). Within that use case, I can control when the client initiates the operation in the subscription's resolver context, but there was no way of knowing when the client has ended the subscription. Considering that, I added an optional asynchronous callback to the initialization of the server called **on_subscription_complete**. 

I've been a heavy user of the [**apollo-server**](https://github.com/apollographql/apollo-server) which is a node equivalent of ariadne. By no means I want to compare, or blame whos "better", but just in the intention of presenting my idea, **apollo-server** has a similar feature (in it's version 3) called **onOperationComplete** that lets the developer know that the operation was ended by the client.

That being said, I made some simple modifications to make this feature possible in ariadne:

1. Created a class called **subscription,** within the **contrib/subscritptions** folder, to represent subscriptions within the **asgi** implementation context. I did it like this to have the **opeartion_name** in the subscriptions so I'm able to control which subscription was closed.

```
from typing import AsyncGenerator

class Subscription:
    def __init__(self, operation_name: str, async_generator: AsyncGenerator):
        self.operation_name = operation_name
        self.async_generator = async_generator
```

2.  Added the **on_subscription_complete** callback to the initialization of the server:

```
OnConnect = Callable[[WebSocket, Any], None]
OnDisconnect = Callable[[WebSocket], None]
OnSubscriptionComplete = Callable[[Subscription], Awaitable]
```

```
class GraphQL:
    def __init__(
        self,
        schema: GraphQLSchema,
        *,
        context_value: Optional[ContextValue] = None,
        root_value: Optional[RootValue] = None,
        on_connect: Optional[OnConnect] = None,
        on_disconnect: Optional[OnDisconnect] = None,
        on_subscription_complete: Optional[OnSubscriptionComplete] = None,
        validation_rules: Optional[ValidationRules] = None,
        debug: bool = False,
        introspection: bool = True,
        logger: Optional[str] = None,
        error_formatter: ErrorFormatter = format_error,
        extensions: Optional[Extensions] = None,
        middleware: Optional[Middlewares] = None,
        keepalive: float = None,
    ):
        self.context_value = context_value
        self.root_value = root_value
        self.on_connect = on_connect
        self.on_disconnect = on_disconnect
        self.on_subscription_complete = on_subscription_complete
        self.validation_rules = validation_rules
        self.debug = debug
        self.introspection = introspection
        self.logger = logger
        self.error_formatter = error_formatter
        self.extensions = extensions
        self.middleware = middleware
        self.keepalive = keepalive
        self.schema = schema
```

3. Modified the implementation of **start_websocket_subscription** to support the subscription class

```
  if not success:
            results = cast(List[dict], results)
            await websocket.send_json(
                {"type": GQL_ERROR, "id": operation_id, "payload": results[0]}
            )
  else:
           results = cast(AsyncGenerator, results)
           subscription = Subscription(data.get("operationName"), results)
           subscriptions[operation_id] = subscription
           asyncio.ensure_future(
               self.observe_async_results(results, operation_id, websocket)
           )
```

4. Changed the implementation of  **websocket_server** to creating the subscriptions Dict supporting the subscription class

```
 subscriptions: Dict[str, Subscription] = {}
        await websocket.accept("graphql-ws")
        try:
            while WebSocketState.DISCONNECTED not in (
                websocket.client_state,
                websocket.application_state,
            ):
                message = await websocket.receive_json()
                await self.handle_websocket_message(message, websocket, subscriptions)
```

5. Modified the implementation of **handle_websocket_message** to support the new subscriptions Dict

```
    async def handle_websocket_message(
        self,
        message: dict,
        websocket: WebSocket,
        subscriptions: Dict[str, Subscription],
    ):
        operation_id = cast(str, message.get("id"))
        message_type = cast(str, message.get("type"))
```

6. Created **close_subscription** that receives a subscription and calls the **on_subscription_complete** callback and calls the s**ubscription.async_generator.aclose()** that was called in the previous implementation

```
 async def close_subscription(self, subscription: Subscription):
        if self.on_subscription_complete:
            await self.on_subscription_complete(subscription)
        await subscription.async_generator.aclose()
```

7. Used the **close_subscription** in the **websocket_server** and **handle_websocket_message** to close the subscriptions accordantly

```
    async def websocket_server(self, websocket: WebSocket) -> None:
        subscriptions: Dict[str, Subscription] = {}
        await websocket.accept("graphql-ws")
        try:
            while WebSocketState.DISCONNECTED not in (
                websocket.client_state,
                websocket.application_state,
            ):
                message = await websocket.receive_json()
                await self.handle_websocket_message(message, websocket, subscriptions)
        except WebSocketDisconnect:
            pass
        finally:
            for subscription in subscriptions.values():
                await self.close_subscription(subscription=subscription)
```

```
     async def handle_websocket_message(
        self,
        message: dict,
        websocket: WebSocket,
        subscriptions: Dict[str, Subscription],
    ):
        operation_id = cast(str, message.get("id"))
        message_type = cast(str, message.get("type"))

        if message_type == GQL_CONNECTION_INIT:
            await self.handle_websocket_connection_init_message(
                message,
                websocket,
            )
        elif message_type == GQL_CONNECTION_TERMINATE:
            await self.handle_websocket_connection_terminate_message(
                websocket,
            )
        elif message_type == GQL_START:
            await self.start_websocket_subscription(
                message.get("payload"), operation_id, websocket, subscriptions
            )
        elif message_type == GQL_STOP:
            if operation_id in subscriptions:
                await self.close_subscription(subscription=subscriptions[operation_id])
                del subscriptions[operation_id]
```

And finally, here's me using it

```
async def on_disconnect(websocket):
    print("client disconnect")


async def on_subscription_complete(subscription):
    print(subscription)


def build_server() -> CORSMiddleware:
    schema = build_schema()
    graphql_server = CORSMiddleware(
        GraphQL(
            schema,
            on_connect=on_connect,
            on_disconnect=on_disconnect,
            on_subscription_complete=on_subscription_complete,
            context_value=context_value,
            debug=True,
        ),
        allow_origins=["*"],
        allow_methods=("GET", "POST", "OPTIONS"),
    )

    return graphql_server
```

Well, I hope this is a feature that helps improve ariadne as a whole and I hope I'm able to contribute in some way. If you guys have questions, please let me know. If you guys think this implementation could be done any better, please, also let me know.

Thanks in advance.

### CODE REVIEW

File: ariadne/asgi.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
     Any,
     AsyncGenerator,
     Callable,
+    Awaitable,
```
Comment (rafalp) [inline ariadne/asgi.py:8]: Should go before `Callable`

File: ariadne/asgi.py [hunk 2]
```diff
@@ -25,6 +26,7 @@
     DATA_TYPE_MULTIPART,
     PLAYGROUND_HTML,
 )
+from .contrib.subscriptions.subscription import Subscription
```
Comment (rafalp) [inline ariadne/asgi.py:29]: Core features shouldn't rely on contrib package. Contrib is for extra opt-in features. This class can be moved to `ariadne.subscriptions`.

File: pyproject.toml [hunk 3]
```diff
@@ -1,23 +0,0 @@
-[tool.black]
```
Comment (rafalp) [inline pyproject.toml:1]: Why delete this file?

**Review by rafalp — CHANGES_REQUESTED** [review]
Looking good to me, but I see you've deleted black config and re-formatted snapshots with it. ;)

File: ariadne/asgi.py (hunk 1 above)
Comment (soares-fernando) [reply ariadne/asgi.py:8]: Ok, no problem, gonna change that.

File: ariadne/asgi.py (hunk 2 above)
Comment (soares-fernando) [reply ariadne/asgi.py:29]: Ok

File: pyproject.toml (hunk 3 above)
Comment (soares-fernando) [reply pyproject.toml:1]: I don't know either, maybe miss click lol.

**Comment by soares-fernando** [comment]
@rafalp I made the modifications as you request. I'm sorry for the delay. Additionally, I added the **context_value** function to the subscription class itself so I'm able to know from which connection the unsubscribe come from. 
Thanks in advance.

File: ariadne/asgi.py [hunk 4]
```diff
@@ -25,6 +26,7 @@
     DATA_TYPE_MULTIPART,
     PLAYGROUND_HTML,
 )
+from ariadne.subscriptions import Subscription
```
Comment (rafalp) [inline ariadne/asgi.py:29]: This should be a relative import between `.logger` and `.types`

File: ariadne/asgi.py [hunk 5]
```diff
@@ -389,7 +392,12 @@ async def start_websocket_subscription(
             )
         else:
             results = cast(AsyncGenerator, results)
-            subscriptions[operation_id] = results
+            subscription = Subscription(
```
Comment (rafalp) [inline ariadne/asgi.py:395]: Do we need a temporary value here? We could do `subscriptions[operation_id] = Subscription(...)`

File: ariadne/subscriptions.py [hunk 6]
```diff
@@ -1,9 +1,9 @@
-from typing import Callable, Dict
+from typing import Callable, Dict, AsyncGenerator
```
Comment (rafalp) [inline ariadne/subscriptions.py:1]:
```suggestion
from typing import AsyncGenerator, Callable, Dict
```

File: ariadne/subscriptions.py [hunk 7]
```diff
@@ -1,9 +1,9 @@
...
 
 from graphql.type import GraphQLSchema
 
 from .objects import ObjectType
-from .types import Subscriber
+from .types import Subscriber, ContextValue
```
Comment (rafalp) [inline ariadne/subscriptions.py:6]:
```suggestion
from .types import ContextValue, Subscriber
```

**Review by rafalp — CHANGES_REQUESTED** [review]
Just some cosmetics

File: ariadne/asgi.py (hunk 4 above)
Comment (soares-fernando) [reply ariadne/asgi.py]: Ok, no problem.

File: ariadne/asgi.py (hunk 5 above)
Comment (soares-fernando) [reply ariadne/asgi.py]: Yeah, I guess not, I changed that.

File: ariadne/subscriptions.py (hunk 6 above)
Comment (soares-fernando) [reply ariadne/subscriptions.py]: Changed.

File: ariadne/asgi.py [hunk 8]
```diff
@@ -25,11 +26,13 @@
     DATA_TYPE_MULTIPART,
     PLAYGROUND_HTML,
 )
+
 from .exceptions import HttpBadRequestError, HttpError
 from .file_uploads import combine_multipart_data
 from .format_error import format_error
 from .graphql import graphql, subscribe
 from .logger import log_error
+from ariadne.subscriptions import Subscription
```
Comment (rafalp) [inline ariadne/asgi.py:35]:
```suggestion
from .subscriptions import Subscription
```

File: ariadne/asgi.py [hunk 9]
```diff
@@ -25,11 +26,13 @@
     DATA_TYPE_MULTIPART,
     PLAYGROUND_HTML,
 )
+
```
Comment (rafalp) [inline ariadne/asgi.py:29]:
```suggestion
```

File: ariadne/asgi.py [hunk 10]
```diff
@@ -307,7 +311,7 @@ async def handle_websocket_message(
             )
         elif message_type == GQL_STOP:
             if operation_id in subscriptions:
-                await subscriptions[operation_id].aclose()
+                await self.close_subscription(subscription=subscriptions[operation_id])
```
Comment (rafalp) [inline ariadne/asgi.py:314]:
```suggestion
                await self.close_subscription(subscriptions[operation_id])
```

File: ariadne/subscriptions.py [hunk 11]
```diff
@@ -1,9 +1,9 @@
-from typing import Callable, Dict
+from typing import AsyncGenerator, Callable, Dict
 
 from graphql.type import GraphQLSchema
 
 from .objects import ObjectType
-from .types import Subscriber
+from .types import Subscriber, ContextValue
```
Comment (rafalp) [inline ariadne/subscriptions.py:6]:
```suggestion
from .types import ContextValue, Subscriber
```

File: ariadne/subscriptions.py [hunk 12]
```diff
@@ -47,3 +47,15 @@ def bind_subscribers_to_graphql_type(self, graphql_type):
...
+class Subscription:
+    def __init__(
+        self,
+        operation_name: str,
+        async_generator: AsyncGenerator,
+        context_value: ContextValue,
+    ):
+        self.operation_name = operation_name
+        self.async_generator = async_generator
+        self.context_value = context_value
```
Comment (rafalp) [inline ariadne/subscriptions.py:61]: Missing blank line at end of file:

```suggestion
        self.context_value = context_value

```

File: ariadne/asgi.py [hunk 13]
```diff
@@ -281,13 +284,13 @@ async def websocket_server(self, websocket: WebSocket) -> None:
             pass
         finally:
             for subscription in subscriptions.values():
-                await subscription.aclose()
+                await self.close_subscription(subscription=subscription)
```
Comment (rafalp) [inline ariadne/asgi.py:287]:
```suggestion
                await self.close_subscription(subscription)
```

File: ariadne/asgi.py [hunk 14]
```diff
@@ -426,3 +433,8 @@ async def observe_async_results(
             websocket.application_state,
         ):
             await websocket.send_json({"type": GQL_COMPLETE, "id": operation_id})
+
+    async def close_subscription(self, subscription: Subscription):
+        if self.on_subscription_complete:
+            await self.on_subscription_complete(subscription)
+        await subscription.async_generator.aclose()
```
Comment (rafalp) [inline ariadne/asgi.py:440]:
```suggestion
        await subscription.async_generator.aclose()

```

**Review by rafalp — APPROVED** [review]

**Comment by rafalp** [comment]
I'm curious why tests aren't running for this PR 🤔

**Comment by soares-fernando** [comment]
@rafalp If anything, let me know. Should I close the PR?

**Comment by rafalp** [comment]
Apologies, this is still on my list to check for including in Ariadne 0.15, but I've just wasn't able to find time for it so far.

🤞I'll get to it sometime this week.

**Comment by soares-fernando** [comment]
> 

No problem @rafalp. I've been using my fork for development, but as soon as ariadne 0.15 is on, I'll be using it. Thanks again.

**Comment by rafalp** [comment]
I'm working on merging this but there's extra work required for this behavior to be fully completed. Currently the implementation assumes that there's single websocket connection per operation, and this is not true. I can have multiple operations per single connection, and I can have operations repeat during same websocket connection. Furthermore not removing closed connections from map is potential memory leak for long-standing connections.

I'm working on addressing those in #834

**Comment by rafalp** [comment]
@soares-fernando I've merged this, but I am not sure we will want to keep it. Apollo Server removed this hook from public API, and I don't see any cleanup logic myself that one would want to do for websocket in this callback that can't be done in `on_disconnect`. Unless developer does something sketchy with mutating the `Subscription` object for extra state?

**Comment by soares-fernando** [comment]
> I'm working on merging this but there's extra work required for this behavior to be fully completed. Currently the implementation assumes that there's single websocket connection per operation, and this is not true. I can have multiple operations per single connection, and I can have operations repeat during same websocket connection. Furthermore not removing closed connections from map is potential memory leak for long-standing connections.
> 
> I'm working on addressing those in #834

Hello @rafalp, how is going?

I'm sorry it took too long for me to respond to this. About the issue you mentioned, in which way does the implementation only consider one operation per connection? I'm looking at the code, and I use operation_id as the identifier for creating new Subscriptions classes, so it should be able to take into account multiple operations per single websocket connection and repeated operations on the same connection as well.

**Comment by soares-fernando** [comment]
> > @soares-fernando I've merged this, but I am not sure we will want to keep it. Apollo Server removed this hook from public API, and I don't see any cleanup logic myself that one would want to do for websocket in this callback that can't be done in `on_disconnect`. Unless developer does something sketchy with mutating the `Subscription` object for extra state?
> 
> I haven't checked the newer version of Apollo Server, I didn't know they had it removed. But as far as my use case goes, I'm using it to reuse subscriptions, or better saying not wasting time and memory having to recreate all the state the leads to the data for the subscription being produced. The bottom line is, why should I recreate all the state the is producing the data for the subscription if its already there?
> 
> This happens because the same user can retrieve the same information in different parts of the screen, and also, when the client unsubscribes, not necessarily it disconnects, so I also need this callback to clean up the subscriptions they are not using at the moment.

I've just now seen that the newest version of Apollo Server doesn't have it anymore, wich for me is a big mistake. I have another application the uses it in a older version of Apollo Server, and there I have subscriptions into to a broadcast type of scenario, where I use redis pub sub pattern to send the same data across several clients. Now, I don't even know how I'm supposed to handle the unsubs to reuse the same pub sub channels. This for me is just a pure mistake, if you look past they had added because people were asking it for the same reason as me, and now they have removed. It just does not make sense to me.

**Comment by rafalp** [comment]
Hey, sorry for tragically delayed response!

Giving those lifecycle hooks another look, I am wondering what new they bring to the table that can't be done using `websocket.scope`dict and managers in `subscribe` functions?

**Pull request merged**